feat(core): support TypeScript 5.5

crisbeto opened the pull request:
Updates the repo to add support for TypeScript 5.5. Includes resolving some compilation errors and broken tests.

crisbeto commented on packages/language-service/testing/src/project.ts:
```diff
@@ -135,11 +135,6 @@ export class Project {
-      // changing since TypeScript will `switchToScriptVersionCache` when a file is opened.
-      // Note that this emulates what we have to do in the server/extension as well.
-      // TODO: remove this once PR #41475 lands
-      this.tsProject.markAsDirty();
```
Context here is that the `markAsDirty` method doesn't exist anymore and the TODO above mentioned that it should've been removed a long time ago.

crisbeto commented on packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:
```diff
@@ -3375,7 +3375,7 @@ runInEachFileSystem((os: string) => {
           env.driveMain();
           const jsContents = env.getContents('test.js');
-          expect(jsContents).toMatch(/function Test_Factory\(t\) { i0\.ɵɵinvalidFactory\(\)/ms);
+          expect(jsContents).toMatch(/function Test_Factory\(t\) { i0\.ɵɵinvalidFactory\(\)/m);
```
The `s` flag wasn't doing anything here so I removed it.

crisbeto commented on packages/compiler-cli/src/ngtsc/reflection/src/type_to_value.ts:
```diff
@@ -83,6 +83,10 @@ export function typeToValue(
         return typeOnlyImport(typeNode, firstDecl);
       }
 
+      if (!ts.isImportDeclaration(firstDecl.parent)) {
```
This is necessary, because import clauses can now be inside `ImportDeclaration` or `JSDocImportTag`.

clydin commented on packages/compiler-cli/test/compliance/test_cases/r3_compiler_compliance/components_and_directives/standalone/GOLDEN_PARTIAL.js:
```diff
@@ -171,7 +171,7 @@ i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "0.0.0-PLACEHOLDE
             type: Directive,
             args: [{
                     standalone: true,
-                    selector: '[direct]',
```
The trailing comma removal is adding a large amount to the diff for this PR.  Is this something that is expected and can we fix that somehow?  Keeping the trailing commas would better align with the formatting throughout the rest of the repo.

crisbeto committed:
feat(core): support TypeScript 5.5

Updates the repo to add support for TypeScript 5.5. Includes resolving some compilation errors and broken tests.

crisbeto commented:
This is still missing approvals for `bazel`, `devtools` and `dev-infra`. Everybody who can approve it for them is away. I'm skipping PullApprove for it, since the changes there are trivial:
- For `bazel` it's a change in the `peerDependencies`.
- For `devltools` it's a fix for an incorrect type.
- For `dev-infra` it's due to the version bump.

pkozlowski-opensource commented:
This PR was merged into the repository by commit e5a6f917225aafa7c5c860f280d2aafe3615727e.
